discovery/ecs/client/aws.go

```diff
@@ -0,0 +1,541 @@
+		Cluster:    cluster.ClusterArn,
+		MaxResults: aws.Int64(maxAPIRes),
+	}
+	log.Debugf("Getting cluster %s services", aws.StringValue(cluster.ClusterName))
+	// Start listing services
+	for {
+		resp, err := c.ecsCli.ListServices(params)
+		if err != nil {
+			return nil, err
+		}
+
+		srvArns := []*string{}
+		for _, srv := range resp.ServiceArns {
+			srvArns = append(srvArns, srv)
+		}
+
+		// Describe services
+		params2 := &ecs.DescribeServicesInput{
+			Cluster:  cluster.ClusterArn,
+			Services: srvArns,
+		}
+
+		resp2, err := c.ecsCli.DescribeServices(params2)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, srv := range resp2.Services {
+			srvs = append(srvs, srv)
+		}
+
+		if resp.NextToken == nil || aws.StringValue(resp.NextToken) == "" {
+			break
+		}
+		params.NextToken = resp.NextToken
+
+	}
+	log.Debugf("Retrieved %d services on cluster %s", len(srvs), aws.StringValue(cluster.ClusterName))
+	return srvs, nil
+}
+
+func (c *AWSRetriever) getTaskDefinitions(tDIDs []*string, useChache bool) ([]*ecs.TaskDefinition, error) {
+	tDefs := []*ecs.TaskDefinition{}
+	var mErr error
+	var wg sync.WaitGroup
+	cached := 0
+	for _, tID := range tDIDs {
+		// If we want cache then check if already present
+		if _, ok := c.cache.getTaskDefinition(aws.StringValue(tID)); useChache && ok {
+			cached++
+			continue
+		}
+		wg.Add(1)
+		go func(tID string) {
+			defer wg.Done()
+			params := &ecs.DescribeTaskDefinitionInput{
+				TaskDefinition: aws.String(tID),
+			}
+			if mErr != nil {
+				return
+			}
+			resp, err := c.ecsCli.DescribeTaskDefinition(params)
+			if err != nil {
+				mErr = err
+				return
+			}
+
+			tDefs = append(tDefs, resp.TaskDefinition)
+		}(*tID)
+	}
+	wg.Wait()
+
+	log.Debugf("Retrieved %d new task definitions, cached %d", len(tDefs), cached)
+	return tDefs, mErr
+}
+
+func (c *AWSRetriever) getInstances(ec2IDs []*string) ([]*ec2.Instance, error) {
+
+	instances := []*ec2.Instance{}
+	log.Debugf("Getting ec2 instances")
+
+	params := &ec2.DescribeInstancesInput{
+		InstanceIds: ec2IDs,
+	}
+
+	// Start describing instances
+	for {
+		resp, err := c.ec2Cli.DescribeInstances(params)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, r := range resp.Reservations {
+			for _, i := range r.Instances {
+				instances = append(instances, i)
+			}
+		}
+
+		if resp.NextToken == nil || aws.StringValue(resp.NextToken) == "" {
+			break
+		}
+		params.NextToken = resp.NextToken
+	}
+
+	log.Debugf("Retrieved %d ec2 instances", len(instances))
+	return instances, nil
+}
+
+// Retrieve will get all the service instance calling multiple AWS APIs
+func (c *AWSRetriever) Retrieve() ([]*types.ServiceInstance, error) {
+
+	// First get the clusters and map them
+	clusters, err := c.getClusters()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cl := range clusters {
+		c.cache.SetCluster(cl)
+	}
+
+	var wg sync.WaitGroup
+	var globErr error
+	globErrMsg := "rrror from other goroutine, stopping AWS SD objects gathering: %s"
```

Comment:
`globErr` is not guarded by a mutex. So the goroutines setting it is a race condition.

I'd recommend using https://godoc.org/golang.org/x/sync/errgroup instead of a trivial wait group for nicer error handling.